### DISCUSSION TITLE
How do you control the starting directory, source code mount location, and mounting behavior of a codespace?

### DISCUSSION THREAD

**Discussion by Nezteb** [community/community#15506] [Codespaces]
For context, I submitted a PR to a repo to add basic codespaces support: https://github.com/livebook-dev/livebook/pull/1138

However, I ran into many issues during the process.

The following information is from https://code.visualstudio.com/docs/remote/devcontainerjson-reference:

> Under `Scenario specific properties > Image or Dockerfile specific properties`
> Property | Type | Description
> -- | -- | --
> workspaceFolder | string | Requires `workspaceMount` be set. Sets the default path that VS Code and other `devcontainer.json` supporting services / tools should open when connecting to the container. Defaults to the automatic source code mount location. ⚠️ Not yet supported in Codespaces or when using Clone Repository in Container Volume.
> 
> Under `Scenario specific properties > Docker Compose specific properties`
> Property | Type | Description
> -- | -- | --
> workspaceFolder | string | Sets the default path that VS Code and other `devcontainer.json` supporting services / tools should open when connecting to the container (which is often the path to a volume mount where the source code can be found in the container). Defaults to `"/"`.

My questions:
1. What is "the automatic source code mount location" for GitHub codespaces? This doesn't seem to be documented anywhere that I can find. Is this location always `/workspaces/<name_of_repo>`? If so, it looks like this isn't configurable?
2. When using Docker Compose for codespaces, no matter what I set `workspaceFolder` to, I always end up in the same `/workspaces/<name_of_repo>` directory. It seems like the property doesn't actually do anything, which is especially weird because VS Code complains if I _don't_ include the property.
    - <img width="322" alt="Screen Shot 2022-04-24 at 1 05 39 PM" src="https://user-images.githubusercontent.com/3588798/164990224-5ccb8ae9-3627-412f-a48c-d66de01488fb.png">
    - Note: `workspaceFolder` _does_ actually work when running this locally with VS Code using `Remote-Containers: Open Folder in container...`, but not in the codespace.

Once in the codespace, if I run `env` I do see a `CODESPACE_VSCODE_FOLDER=/workspaces/<name_of_repo>` variable, but attempting to change it in the `devcontainer.json` as shown below doesn't work:
```
"remoteEnv": {
  "CODESPACE_VSCODE_FOLDER": "/root",
}
```
This `devcontainer.json`:
<img width="1072" alt="Screen Shot 2022-04-24 at 1 19 12 PM" src="https://user-images.githubusercontent.com/3588798/164990887-cd7b2aac-f091-4b2f-be70-825045dd64e7.png">

Results in this `env`:
<img width="1223" alt="Screen Shot 2022-04-24 at 1 19 23 PM" src="https://user-images.githubusercontent.com/3588798/164990995-d89431d6-1b2a-41de-890c-36862c86f857.png">

I've also noticed that if my Dockerfile includes a final `RUN` command that builds/compiles my entire app (which works locally), the codespace seems to overwrite that entire directory when mounting the source code, effectively requiring any user of the codespace to rebuild/recompile the app. There is no way to let the codespace user just start a codespace, use a run command, and have the app start. They'll have to build/compile it themselves first.

Running locally:
<img width="1669" alt="Screen Shot 2022-04-24 at 1 45 49 PM" src="https://user-images.githubusercontent.com/3588798/164991712-9f353625-8fe9-469a-af55-aea3fc0e3462.png">

Versus running in a codespace:
<img width="816" alt="Screen Shot 2022-04-24 at 1 43 11 PM" src="https://user-images.githubusercontent.com/3588798/164991731-f55ca1cf-f2d2-43f9-b789-4d804cb3f661.png">

If my observations are correct, any Dockerfile that does something along the lines of `COPY . .` will result in a bunch of duplicated code. If I'm copying an entire `src/` directory from the host to `/app` in the container's Dockerfile, then Codespaces is putting all of that exact same code in `/workspaces/<name_of_repo>`. Is there no way to prevent this? Do I have to choose between not doing the `COPY` and not duplicating code in the final container? And do I not get a choice of where that code is mounted?

# **tl;dr**
How can I do the following?
1. Control the directory that my codespace starts in (instead of `/workspaces/<name_of_repo>`.
2. Control where the repo code gets mounted into the codespace.
3. Prevent a codespace from overwriting any built/compiled artifacts from the original Docker image when mounting the source code.

**Comment by mraarone** [comment]
I just googled "CODESPACE_VSCODE_FOLDER" and there was _1_ result... this issue page. I, too, am trying to figure out exactly what you've noted, above. Have you made any progress since April? 

If I can overcome this replication issue, I can get python virtual environments working. It seems like virtual environments wouldn't be necessary in a container, but there are some cases where virtual environments are expected, like with security scans.

**Reply by Nezteb** [reply]
Unfortunately no. I ended up giving up on Codespaces in general as a product, especially because of their weird pricing model: https://github.com/community/community/discussions/15645

EDIT: It turns out GitHub will be changing the pricing model to something way more sustainable. That said, I still never got a solution for the `CODESPACE_VSCODE_FOLDER` thing...

**Reply by TommyLCox** [reply]
I can't believe this remained unanswered. One can only assume - there is no answer.

I haven't tried it, but I am thinking on creating a repository that is really empty except for having a codespace. Then, in that code space, make a folder and clone the repository(ies) that i want into that specific subfolder. Also, there is https://github.blog/2022-04-20-codespaces-multi-repository-monorepo-scenarios/ .. I am going to try using that for cloning the multiple repositories.

**Comment by julius-gummersbach** [comment]
> How can I do the following?
> 1. Control the directory that my codespace starts in (instead of `/workspaces/<name_of_repo>`.
> 2. Control where the repo code gets mounted into the codespace.
> 3. Prevent a codespace from overwriting any built/compiled artifacts from the original Docker image when mounting the source code.  

For `1` I just set the `workspaceFolder` field in my `devcontainer.json` file accordingly. It worked for me. Seems like they updated something since you last tried it. Maybe give it another shot. The repository code still gets mounted to `/workspaces/<repo-name>` , but vscode starts in the specified directory.  

As a workaround for `3` I cloned and build the code in my Dockerfile to a location different from `/workspaces/<repo-name>` ( just `/<repo-name` in my case ). It does not get overwritten there and with `1` I can just let vscode start in `/<repo-name>` with all my artifacts already build. Far from ideal because there is duplicate code sitting in `/workspaces/<repo-name>`, but it works ¯\_(ツ)_/¯


`.devcontainer/devcontainer.json` :
```
{
	"name": "Foo",
	"build": { "dockerfile": "Dockerfile" },
	"workspaceFolder": "/foo"
}
```

`.devcontainer/Dockerfile` :
```
FROM ubuntu

# [...]

RUN git clone <url> foo
WORKDIR "/foo"
RUN mvn clean install #(in my case)
```

Hope I could help!
Cheers, Julius

**Comment by weihongliang233** [comment]
It's 2024 now.
I still encounterd the same problem.
![image](https://github.com/community/community/assets/52309635/49bd6d12-f378-4aa3-8169-f5671e59ae83)

**Comment by wmarshall484** [comment]
Also running into this issue.
![image](https://github.com/user-attachments/assets/0d27fe67-4e43-4742-bd6a-9f803ad586ca)

**Comment by SAYAF-14** [comment]
Response #2

**Comment by rady-msaad** [comment]
Was there progress or solution for this. I have encountered same problem and was able to change workspace folder using environment variable in docker-compose which shows correctly in codespace but still the starting directory is reset to codespace repository

**Reply by meekrob** [reply]
Looking for an answer in 2025...

**Comment by leidix** [comment]
+1

**Comment by ChipWolf** [comment]
+1
